File: src/Algorithms/Moor'e_majority_voting_algorithm/169_Majority_Element.java
class Solution {
    public int majorityElement(int[] nums) {
        int majorityEl = -1;
        int votes = 0;

        for (int num : nums) {
            if (votes <= 0) {
                majorityEl = num;
            }

            if (num == majorityEl) {
                votes++;
            } else {
                votes--;
            }
        }

        votes = 0;
        for (int num : nums) {
            if (num == majorityEl) {
                votes++;
            }
        }

        if (votes >= (nums.length / 2))
            return majorityEl;
        return -1;
    }
}

File: src/Algorithms/Moore_majority_voting_algorithm/169_Majority_Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */


var majorityElement = function(nums) {
    let majorityEl = -1;
    let votes = 0;

    nums.forEach(num => {
        if (votes <= 0) {
            majorityEl = num;
        }

        if (num === majorityEl) {
            votes += 1;
        } else {
            votes -= 1;
        }
    });

    votes = 0;
    nums.forEach(num => {
        if (num === majorityEl) {
            votes += 1;
        }
    });

    if (votes >= Math.floor(nums.length / 2)) return majorityEl;
    return -1;
};